kms/capi/ncrypt_windows.go

```diff
@@ -0,0 +1,428 @@
+	CRYPT_ACQUIRE_ALLOW_NCRYPT_KEY_FLAG  = uint32(0x00010000)
+	CRYPT_ACQUIRE_PREFER_NCRYPT_KEY_FLAG = uint32(0x00020000)
+	CRYPT_ACQUIRE_ONLY_NCRYPT_KEY_FLAG   = uint32(0x00040000)
+
+	AT_KEYEXCHANGE = uint32(1)
+	AT_SIGNATURE   = uint32(2)
+
+	// Legacy CryptoAPI flags
+	bCryptPadPKCS1 = uint32(2)
+
+	// Magic numbers for public key blobs.
+	rsa1Magic = 0x31415352 // "RSA1" BCRYPT_RSAPUBLIC_MAGIC
+	ecs1Magic = 0x31534345 // "ECS1" BCRYPT_ECDSA_PUBLIC_P256_MAGIC
+	ecs3Magic = 0x33534345 // "ECS3" BCRYPT_ECDSA_PUBLIC_P384_MAGIC
+	ecs5Magic = 0x35534345 // "ECS5" BCRYPT_ECDSA_PUBLIC_P521_MAGIC
+)
+
+var (
+	// curveNames maps bcrypt.h curve names to elliptic curves.
+	curveNames = map[string]elliptic.Curve{
+		"nistP256": elliptic.P256(), // BCRYPT_ECC_CURVE_NISTP256
+		"nistP384": elliptic.P384(), // BCRYPT_ECC_CURVE_NISTP384
+		"nistP521": elliptic.P521(), // BCRYPT_ECC_CURVE_NISTP521
+	}
+
+	// algIDs maps crypto.Hash values to bcrypt.h constants.
+	hashAlgorithms = map[crypto.Hash]string{
+		crypto.SHA1:   "SHA1",   // BCRYPT_SHA1_ALGORITHM
+		crypto.SHA256: "SHA256", // BCRYPT_SHA256_ALGORITHM
+		crypto.SHA384: "SHA384", // BCRYPT_SHA384_ALGORITHM
+		crypto.SHA512: "SHA512", // BCRYPT_SHA512_ALGORITHM
+	}
+
+	crypt32 = windows.MustLoadDLL("crypt32.dll")
+	nCrypt  = windows.MustLoadDLL("ncrypt.dll")
+
+	certFindCertificateInStore      = crypt32.MustFindProc("CertFindCertificateInStore")
+	cryptFindCertificateKeyProvInfo = crypt32.MustFindProc("CryptFindCertificateKeyProvInfo")
+	nCryptCreatePersistedKey        = nCrypt.MustFindProc("NCryptCreatePersistedKey")
+	nCryptExportKey                 = nCrypt.MustFindProc("NCryptExportKey")
+	nCryptFinalizeKey               = nCrypt.MustFindProc("NCryptFinalizeKey")
+	nCryptFreeObject                = nCrypt.MustFindProc("NCryptFreeObject")
+	nCryptOpenKey                   = nCrypt.MustFindProc("NCryptOpenKey")
+	nCryptOpenStorageProvider       = nCrypt.MustFindProc("NCryptOpenStorageProvider")
+	nCryptGetProperty               = nCrypt.MustFindProc("NCryptGetProperty")
+	nCryptSetProperty               = nCrypt.MustFindProc("NCryptSetProperty")
+	nCryptSignHash                  = nCrypt.MustFindProc("NCryptSignHash")
+)
+
+type BCRYPT_PKCS1_PADDING_INFO struct {
+	pszAlgID *uint16
+}
+
+func errNoToStr(e uint32) string {
+	switch e {
+	case NTE_INVALID_PARAMETER:
+		return "NTE_INVALID_PARAMETER"
+	case NTE_BAD_FLAGS:
+		return "NTE_BAD_FLAGS"
+	case NTE_BAD_KEYSET:
+		return "NTE_BAD_KEYSET"
+	case NTE_NO_MORE_ITEMS:
+		return "NTE_NO_MORE_ITEMS"
+	case NTE_NOT_SUPPORTED:
+		return "NTE_NOT_SUPPORTED"
+	case SCARD_W_CANCELLED_BY_USER:
+		return "User cancelled smartcard action"
+	default:
+		return fmt.Sprintf("0x%X", e)
+	}
+}
+
+// wide returns a pointer to a a uint16 representing the equivalent
+// to a Windows LPCWSTR.
+func wide(s string) *uint16 {
+	w, _ := syscall.UTF16PtrFromString(s)
+	return w
+}
+
+func NCryptOpenStorage(provider string) (uintptr, error) {
```

Comment:
All or most of these methods should be private for now unless you are using this package differently.